src/structures/paging/mapper/mod.rs

```diff
@@ -275,11 +318,172 @@ pub trait Mapper<S: PageSize> {
         Self: Sized,
         A: FrameAllocator<Size4KiB> + ?Sized;
 
+    /// Maps the given range of frames to the range of virtual pages.
+    ///
+    /// ## Safety
+    ///
+    /// This is a convencience function that invokes [`Mapper::map_to_with_table_flags`] internally, so
+    /// all safety requirements of it also apply for this function.
+    ///
+    /// ## Panics
+    ///
+    /// This function panics if the amount of pages does not equal the amount of frames.
+    ///
+    /// ## Errors
+    ///
+    /// If an error occurs half-way through a [`MapperFlushRange<S>`] is returned that contains the frames that were successfully mapped.
+    unsafe fn map_to_range_with_table_flags<A>(
+        &mut self,
+        pages: PageRange<S>,
+        frames: PhysFrameRange<S>,
+        flags: PageTableFlags,
+        parent_table_flags: PageTableFlags,
+        frame_allocator: &mut A,
+    ) -> Result<MapperFlushRange<S>, (MapToError<S>, MapperFlushRange<S>)>
+    where
+        Self: Sized,
+        A: FrameAllocator<Size4KiB> + ?Sized,
+    {
+        assert_eq!(pages.count(), frames.count());
+
+        pages
+            .zip(frames)
+            .try_for_each(|(page, frame)| {
+                self.map_to_with_table_flags(
+                    page,
+                    frame,
+                    flags,
+                    parent_table_flags,
+                    frame_allocator,
+                )
+                .map(|_| ())
+                .map_err(|e| {
+                    (
+                        e,
+                        MapperFlushRange::new(PageRange {
+                            start: pages.start,
+                            end: page,
+                        }),
+                    )
+                })
+            })
+            .map(|_| MapperFlushRange::new(pages))
+    }
+
+    /// Maps frames from the allocator to the given range of virtual pages.
+    ///
+    /// ## Errors
+    ///
+    /// If an error occurs half-way through a [`MapperFlushRange<S>`] is returned that contains the frames that were successfully mapped.
+    fn map_range_with_table_flags<A>(
```

Comment:
Okay, so because `FrameAllocator` guarantees that all allocated pages are unused most of these don't apply. The only exception I see is the comment about the `GLOBAL`, that should still apply with new pages. Are there any other caller guarantees that I should mention in the safety comment?

Reply:
I think the second point about uninitialized values with invalid layout applies here too. The `unmap` function is safe (as it can only cause a page fault, not violate memory safety), so the following sequence would still be problematic:

- map a page
- write something to the page
- unmap the page (this is a safe operation)
- remap the page to a new physical frame using e.g. `map_range_with_table_flags`

Now we basically replaced the values we've written before with some arbitrary data, which can lead to undefined behavior.

The safety comment could just be the _"the safety requirements of `map_to` apply here as well"_ statement we already use in other places (with the exception of the _"aliasing of &mut references"_ point).

Reply:
Oh right I completely forgot about that